apps/dashboard/src/lib/validators/uptime-module.ts

```diff
@@ -0,0 +1,87 @@
+import { BehaviorSubject } from 'rxjs'
+import {
+  calculateApy,
+  getValidatorUptimeSinceDate,
+  uptimePeriodDefinition,
+  type UptimeValue
+} from '@api/utils/entities/component/validator'
+import BigNumber from 'bignumber.js'
+import type { ValidatorCollectionItem } from '@common/gateway-sdk'
+
+type ValidatorAddress = string
+
+export type UptimeData = Partial<
+  Record<UptimeValue, Record<ValidatorAddress, number | undefined>>
+>
+
+export const UptimeModule = () => {
+  const isLoading = new BehaviorSubject<boolean>(false)
+  let validators:
+    | Promise<{ address: string; totalStakeInXRD: BigNumber }[]>
+    | undefined
+  let uptimeData: UptimeData = {}
+  let totalAmountStaked: BigNumber | undefined
+  return {
+    setValidators: (
+      promise: Promise<{ address: string; totalStakeInXRD: BigNumber }[]>
+    ) => {
+      validators = promise
+      promise.then((v) => {
+        totalAmountStaked = v.reduce(
+          (prev, cur) => prev.plus(cur.totalStakeInXRD),
+          new BigNumber(0)
+        )
+      })
+    },
+    getDataForUptime: (uptime: UptimeValue) => uptimeData[uptime] || {},
+    getApy: (validator: ValidatorCollectionItem, uptime: UptimeValue) => {
+      if (!totalAmountStaked) {
+        throw new Error('Total amount staked not set')
+      }
+      const address = validator.address
+      const fee = Number(validator.effective_fee_factor?.current?.fee_factor)
+      const uptimePercentage = uptimeData[uptime]?.[address]
+      return calculateApy(fee, uptimePercentage, totalAmountStaked)
+    },
+    maybeQueryUptime: async (uptime: UptimeValue | undefined) => {
+      if (!uptime) return
+      if (!validators) {
+        throw new Error('Validators not set')
+      }
+      if (uptimeData[uptime]) {
+        return uptimeData[uptime]
+      }
+
+      uptimeData[uptime] = {}
+
+      isLoading.next(true)
+
+      return validators
+        .then((v) =>
+          getValidatorUptimeSinceDate(v.map((v) => v.address))(
+            uptimePeriodDefinition[uptime].getStartingPoint()
+          )
+        )
+        .then((value) => {
+          isLoading.next(false)
+          if (value.isOk()) {
+            uptimeData[uptime] = value.value
+            return uptimeData[uptime]
+          }
+        })
+        .catch((e) => {
+          console.error(e)
+          isLoading.next(false)
+          uptimeData[uptime] = undefined
+        })
+    },
+    clean: () => {
+      validators = undefined
+      uptimeData = {}
+      totalAmountStaked = undefined
+    },
+    isLoading$: isLoading.asObservable()
+  }
+}
+
+export const uptimeModule = UptimeModule()
```

Comment:
instead of a function - can we put the state at the top level in this file, so all importers share the state?

Reply:
Hm, by state you mean `uptimeData`? Currently it's encapsulated and shared between all importers who import `uptimeModule` directly so I do not see a reason why we would want to put it globally? 🤔 To be honest I think I'm going to add unit test for that module so having state inside makes it also easier to test

Reply:
yeah I figured it was shared - so I just thought why not just put the state in the module instead of in a function scope.

You can still test it this way

Reply:
For example the way you're proposing I can't really run multiple test in parallel because if I did they'd all share same data source (which might be undesired)

Reply:
On top of that (because I just started writing some tests) - having it inside function allows me to mock some dependencies much more easily

Reply:
I don't see an issue with `UptimeModule` being encapsulated in a function